src/torrent_info.cpp

```diff
@@ -579,24 +580,30 @@ namespace {
 				return false;
 			}
 
-			std::string path = root_dir;
-			sanitize_append_path_element(path, e.first);
+			string_view filename = { e.first.data() + info_ptr_diff, e.first.size() };
+			while (!filename.empty() && filename.front() == TORRENT_SEPARATOR)
+				filename.remove_prefix(1);
+
+			bool const leaf_node = e.second.dict_size() == 1 && e.second.dict_at(0).first.empty();
+			bool const single_file = leaf_node && top_level && tree.dict_size() == 1;
 
-			if (e.second.dict_size() == 1 && e.second.dict_at(0).first.empty())
+			std::string path = single_file ? std::string() : root_dir;
+			sanitize_append_path_element(path, filename);
+
+			if (leaf_node)
 			{
-				if (single_file && tree.dict_size() == 1)
+				if (filename.size() > path.length()
+					|| path.substr(path.size() - filename.size()) != filename)
 				{
-					// this is a single file torrent
-					// drop the torrent name from the path to match the bahavior
-					// of v1 metadata
-					path.clear();
-					sanitize_append_path_element(path, e.first);
+					// if the filename was sanitized and differ, clear it to just use path
+					filename = {};
```

Comment:
like the v1 add-file path, `filename` has been appended to `path`. If the ending of `path` is different from `filename`, it means it was sanitized in some way, and we should use the full path instead of this filename, so clear it. (this behavior is a bit unintuitive of `add_file_borrow()`)